safenode/src/network/command.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{
+    error::Error,
+    safe_msg::{SafeRequest, SafeResponse},
+    EventLoop,
+};
+use crate::network::error::Result;
+use futures::channel::oneshot;
+use libp2p::{multiaddr::Protocol, request_response::ResponseChannel, Multiaddr, PeerId};
+use std::collections::{hash_map, HashSet};
+use tracing::warn;
+use xor_name::XorName;
+
+/// Commands to send to the Swarm
+#[derive(Debug)]
+pub(crate) enum CmdToSwarm {
+    StartListening {
+        addr: Multiaddr,
+        sender: oneshot::Sender<Result<()>>,
+    },
+    Dial {
+        peer_id: PeerId,
+        peer_addr: Multiaddr,
+        sender: oneshot::Sender<Result<()>>,
+    },
+    StoreData {
+        xor_name: XorName,
+        sender: oneshot::Sender<Result<()>>,
+    },
+    GetDataProviders {
+        xor_name: XorName,
+        sender: oneshot::Sender<HashSet<PeerId>>,
+    },
+    SendSafeRequest {
+        req: SafeRequest,
+        peer: PeerId,
+        sender: oneshot::Sender<Result<SafeResponse>>,
+    },
+    SendSafeResponse {
+        resp: SafeResponse,
+        channel: ResponseChannel<SafeResponse>,
+    },
+}
+
+impl EventLoop {
+    pub(crate) fn handle_command(&mut self, command: CmdToSwarm) -> Result<(), Error> {
+        match command {
+            CmdToSwarm::StartListening { addr, sender } => {
+                let _ = match self.swarm.listen_on(addr) {
+                    Ok(_) => sender.send(Ok(())),
+                    Err(e) => sender.send(Err(e.into())),
+                };
+            }
+            CmdToSwarm::Dial {
+                peer_id,
+                peer_addr,
+                sender,
+            } => {
+                if let hash_map::Entry::Vacant(e) = self.pending_dial.entry(peer_id) {
+                    let _routing_update = self
+                        .swarm
+                        .behaviour_mut()
+                        .kademlia
+                        .add_address(&peer_id, peer_addr.clone());
```

Comment:
This is interesting. Why do we add the address before a dial? Asking juut to understand better what's happening.

Reply:
I think `add_address` just inserts the peer to the DHT, but does not actually connect to it. While `dial` connects to the peer. I've observed this pattern through the logs while fetching the chunks.